qiskit_nature/circuit/library/ansaetze/ucc.py

```diff
@@ -0,0 +1,173 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+TODO.
+"""
+
+from functools import partial
+from typing import Callable, List, Optional, Tuple, Union
+
+import logging
+
+from qiskit_nature.components.variational_forms.excitation_op_builder import ExcitationOpBuilder
+from .evolved_operator_ansatz import EvolvedOperatorAnsatz
+
+logger = logging.getLogger(__name__)
+
+
+# TODO: Added to pass mypy, need change to real class
+class QubitOpConverter:
+    """ The QubitOpConverter """
+
+    def convert(self, op):
+        """TODO."""
+        pass
+
+
+class UCC(EvolvedOperatorAnsatz):
+    """The Unitary Coupled-Cluster Ansatz."""
+
+    EXCITATION_TYPE = {
+        's': 1,
+        'd': 2,
+        't': 3,
+        'q': 4,
+    }
+
+    # TODO: expose evolution and reps here, too?
+    def __init__(self, qubit_op_converter: QubitOpConverter,
```

Comment:
Convertor is needed since the excitation operators are created as Fermionic and they need to be mapped consistently with the the main Hamiltonian operator for which this is the state. Whether it could be optional here I guess is choice over how to set this up - since num_particles and num_spin_orbitals seems to be allowed to be Optional here maybe the convertor can be in that regard too. Ultimately before use they need to be set of course.